File: python_backend/game_logic/ranges/get_range.py
from game_logic.ranges.paths.get_paths import get_paths
from game_logic.ranges.jumps.get_jumps import get_jumps
from game_logic.printers.print_board import print_board
from game_logic.test_objects.get_standard_range_defs import get_standard_range_defs
from game_logic.test_objects.get_standard_id_dict import get_standard_id_dict
from game_logic.test_objects.sample_board_dicts import sample_board_dicts


def get_range(id_dict, range_defs, board, sqr, color):
    """ """
    id_ = board[sqr]
    fen_id = id_[1].lower()
    name = id_dict[fen_id]
    def_ = range_defs[name]
    offsets = def_[color]['offsets']
    step_funcs = def_[color]['spans']
    paths = get_paths(step_funcs, board, sqr, color)
    jumps = get_jumps(offsets, board, sqr, color)
    paths.extend(jumps)
    return paths


if __name__ == "__main__":

    board = sample_board_dicts['ranges_test']
    ranges = \
        {'BB1': [],
         'BB2': [],
         'BK1': [],
         'BN1': [],
         'BN2': [],
         'BQ1': [],
         'BR1': [],
         'BR2': [],
         'WB1': [],
         'WB2': [],
         'WB3': [],
         'WK1': [],
         'WN1': [],
         'WN2': [],
         'WQ1': [],
         'WR1': [],
         'WR2': [],
         'WR3': [],
         'WR4': []}
    id_defs = get_standard_id_dict()
    range_defs = get_standard_range_defs()
    piece_range = get_range(id_defs, range_defs, board, (3, 3), 'W')
    print_board(board, heading="c3", highlights=piece_range)
    piece_range = get_range(id_defs, range_defs, board, (4, 5), 'W')
    print_board(board, heading="d5", highlights=piece_range)

File: node.js/ranges/getRange.js
const  getPaths  = require("./paths/getPaths");
const  getJumps  = require("./jumps/getJumps");
const  printBoard  = require("../printers/printBoard");
const  getStandardPieceDefs  = require("../testObjects/getStandardPieceDefs");
const  getStandardIdDict  = require("../testObjects/getStandardIdDict");
const  sampleBoardDicts  = require("../testObjects/sampleBoardDicts");
const mapListXyToRf = require("../coordType/mapListXyToRf");
const { xyToRf } = require("../coordType/crdCnvrt");


function getRange(board, sqr, color, pieceDefs, idDict) {
    /* **/
    var id = board[xyToRf(...sqr)]
    var fenId = id[1].toLowerCase()
    var name = idDict[fenId]
    var def = pieceDefs[name]
    var offsets = def[color]['offsets']
    var stepFuncs = def[color]['spans']
    var paths = getPaths(stepFuncs, board, sqr, color)
    var jumps = getJumps(offsets, board, sqr, color)
    paths.push(...jumps)
    return paths;
}

module.exports = getRange;

if (require.main === module) {
    var board = sampleBoardDicts['ranges_test']
    var idDict = getStandardIdDict()
    var pieceDefs = getStandardPieceDefs()
    var pieceRange;
    for (var sqr of [[2, 1], [3, 1], [4, 1], [1, 3], [2, 3], [3, 3], [4, 3], [6, 3], [7, 3], [8, 3]]) {
        pieceRange = getRange(board, sqr, 'W', pieceDefs, idDict)
        printBoard(board, `\n${board[xyToRf(...sqr)]}`, mapListXyToRf(pieceRange), null, [xyToRf(...sqr)])
    }
    for (var sqr of [[1, 5], [2, 5], [3, 5], [4, 5], [6, 5], [7, 5], [8, 5]]) {
        pieceRange = getRange(board, sqr, 'B', pieceDefs, idDict)
        printBoard(board, `\n${board[xyToRf(...sqr)]}`, mapListXyToRf(pieceRange), null, [xyToRf(...sqr)])
    }
}

